compiler/rustc_lint/src/builtin.rs

```diff
@@ -1362,66 +1367,82 @@ impl UnreachablePub {
         cx: &LateContext<'_>,
         what: &str,
         def_id: LocalDefId,
-        vis: &hir::Visibility<'_>,
         span: Span,
+        vis_span: Span,
         exportable: bool,
     ) {
         let mut applicability = Applicability::MachineApplicable;
-        match vis.node {
-            hir::VisibilityKind::Public if !cx.access_levels.is_reachable(def_id) => {
```

Comment:
Not sure why the "is public" condition was moved out of `perform_lint`, it's still checked on every use of `perform_lint` below.